### PR TITLE
Module 4 - Project Exercise: Part 5.8 - Start the app! todoapp.servic…

### PR DESCRIPTION
**Pull request by abdulraoufatia**
**Start the app**
You should now be able to run the app on the host.

You use `poetry run flask run` when running the app locally on port 5000, but on the VM you will
need an extra option to run it correctly. Set the "host" to "0.0.0.0" in order to receive external
requests. Use `poetry run flask run --help` for documentation, or see their online documentation.

If all the previous Ansible tasks ran successfully, you should be able to manually SSH onto the
host, "cd" to the correct directory, start up the app and then visit the site in your browser using
the IP address of the host VM followed by ":5000".

But to run the app via Ansible, we will not run the command directly. First of all, it needs to be
run as a background process that doesn't stop when Ansible closes its connection. Secondly,
we need to be able to stop and start it in the future, for example restarting it when we have a
new version to run.

The correct way to run it is via a tool called systemd, which is built into Linux.
Create a file called "todoapp.service", to define your app as a systemd service. It should look
like the following:

```
[Unit]
Description=My To-Do App
[Service]
User=ec2-user
WorkingDirectory=/opt/todoapp
ExecStart=startup command goes here, including host
```

You may need the absolute path to Poetry, e.g. "/home/ec2-user/.local/bin/poetry run flask
run ..."

Add, commit and push this file to your Git repositor

### CODE REVIEW

File: todoapp.service [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[Unit]
+Description=My To-Do App
+
+[Service]
+User=ec2-user
+WorkingDirectory=/home/ec2-user/.local/bin/poetry
```
Comment (hugh-emerson) [inline todoapp.service:6]: When you run the todo app on your own computer, what do you do?

You run the command "poetry run flask run". And you run it from your project folder, the one that contains pyproject.toml etc and a todo_app subfolder with all your python files. The same place that you run "poetry install" from.

You should do the same thing on the remote VM. The "WorkingDirectory" is the directory (folder) you want to work in. There is no todo app inside the /home/ec2-user/.local/bin folder. So "poetry run flask run" will not be able to do anything. You want `WorkingDirectory=/opt/todoapp` as the PDF gave you.

I'm assuming your playbook is correct, but you haven't shared it. So I'm assuming that you are successfully putting your project files into /opt/todoapp.

**Comment by hugh-emerson** [comment]
It's possible that this file will work if your PATH is updated appropriately. You still haven't shared your ansible playbook and you haven't shared what error you are facing (if any) so I am guessing here, but it's very likely that you will get an error when using this file.

See the end of my previous comment:
> The issue that maybe you were trying to resolve is getting it to recognise the "poetry" command. Since it can't find the "poetry" program, you can provide the full file path. This was also given to you in the PDF:
> /home/ec2-user/.local/bin/poetry run flask run --host 0.0.0.0

The idea is instead of trying to run the command "poetry", hoping that it will be found, you should use the full filepath, "/home/ec2-user/.local/bin/poetry"

Please resubmit in Aptem once your submission is **complete** - I expect a playbook, inventory, and .env.j2 file in addition to todoapp.service. If you have any issues or questions, please reach out to us via email or Slack, describing what you would like help with.

**Comment by abdulraoufatia** [comment]
> It's possible that this file will work if your PATH is updated appropriately. You still haven't shared your ansible playbook and you haven't shared what error you are facing (if any) so I am guessing here, but it's very likely that you will get an error when using this file.
> 
> See the end of my previous comment:
> 
> > The issue that maybe you were trying to resolve is getting it to recognise the "poetry" command. Since it can't find the "poetry" program, you can provide the full file path. This was also given to you in the PDF:
> > /home/ec2-user/.local/bin/poetry run flask run --host 0.0.0.0
> 
> The idea is instead of trying to run the command "poetry", hoping that it will be found, you should use the full filepath, "/home/ec2-user/.local/bin/poetry"
> 
> Please resubmit in Aptem once your submission is **complete** - I expect a playbook, inventory, and .env.j2 file in addition to todoapp.service. If you have any issues or questions, please reach out to us via email or Slack, describing what you would like help with.

I have actioned the following:

- playbook
- env.j2
- ini file
- todoapp.service

**Review by hugh-emerson — APPROVED** [review]
Good work - these files all look correct. I've left two minor comments and you can merge this when you're ready.

However, your app still returned an error. Did you try testing it? The aim of the exercise was to run your app through Ansible. i.e. you should be able to visit  http://35.179.55.173:5000/ (the host VM's address) and see your todo app running.
But when I visited that address all I see is an error. In future please test your work before submitting it for review and reach out if you have any issues.

However, in this case, the actual problem was very small so I'll still approve the submission in Aptem. The issue was the .env file was filled out incorrectly when you entered values for the key/token. I've taken down the VMs now, so no need to try fixing it.

Final note - you should keep the README.md up to date with useful info. For this exercise, that means documenting that these files exist and what command the reader should run to make use of them.

**Pull request merged**